hippo.go

```diff
@@ -126,80 +143,128 @@ func (c *Client) Do(ctx context.Context, req *http.Request) ([]byte, error) {
 	return buf, nil
 }
 
-func (c *Client) EncodeTagBatchPart(rFull *TagBatchPart) ([][]byte, int, error) {
+// SendBatchPart sends a batch to the server, in multiple requests, if necessary.
+// - may return non-nil SendBatchResult on error, it'll report how far we got
+// - error will contain SendBatchResult info
+func (c *Client) SendBatch(ctx context.Context, url string, batch *TagBatchPart) (*SendBatchResult, error) {
 	c.lock.RLock()
-	defer c.lock.RUnlock()
+	sender := c.Sender
+	c.lock.RUnlock()
+
+	// compact the batch, grouping the same values together
+	batch = compactTagBatchPart(*batch)
+	batch.Sender = sender
+
+	batchParts, err := c.split(batch)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := &SendBatchResult{
+		PartsTotal:   len(batchParts),
+		UpsertsTotal: len(batch.Upserts),
+		DeletesTotal: len(batch.Deletes),
+		BatchGUID:    "", // not known until we send the first part
+	}
+
+	for i, batchPart := range batchParts {
+		batchPart.BatchGUID = ret.BatchGUID
+
+		requestBytes, err := json.Marshal(batchPart)
+		if err != nil {
+			return ret, fmt.Errorf("Error JSON marshalling request batch - [%s] - underlying error: %s", ret, err)
+		}
+
+		req, err := c.NewTagBatchPartRequest("POST", url, requestBytes)
+		if err != nil {
+			return ret, fmt.Errorf("Error building request to %s - [%s] - underlying error: %s", url, ret, err)
+		}
+		responseBytes, err := c.Do(ctx, req)
+		if err != nil {
+			return ret, fmt.Errorf("Error POSTing populators to %s - [%s] - underlying error: %s", url, ret, err)
+		}
+
+		if i == 0 {
+			// first response returns the batch GUID, which we need to include in subsequent batches
+			apiResponse := APIServerResponse{}
+			if err := json.Unmarshal(responseBytes, &apiResponse); err != nil {
+				return ret, fmt.Errorf("Error unmarshalling API batch response - [%s] - underlying error: %s", ret, err)
+			}
+			if apiResponse.GUID == "" {
+				return ret, fmt.Errorf("API response did not include a GUID for subsequent batches - [%s] - underlying error: %s", ret, apiResponse.Error)
+			}
```

Comment:
Should we do any additional logging/handling if the `apiResponse` contains an error? May be useful to somehow identify/know that part of the operation had failed for this client(on the server's side) if sending in multiple batches.

Reply:
I just updated to handle when `Error` isn't empty string, and updated the returned error to have more context:
https://github.com/kentik/gohippo/blob/e124df8e793b6d063cd41ef7151105db19f393aa/hippo.go#L193-L198

https://github.com/kentik/gohippo/commit/e124df8e793b6d063cd41ef7151105db19f393aa

Note: these conditions won't actually be reached today, since any time it's not a success, the status code isn't 200. When that's the case, this is the error: https://github.com/kentik/gohippo/blob/e124df8e793b6d063cd41ef7151105db19f393aa/hippo.go#L184
which looks like this
https://github.com/kentik/gohippo/blob/e124df8e793b6d063cd41ef7151105db19f393aa/hippo_test.go#L502

That error message will contain the full server response, which will likely be that same JSON.